### ISSUE TITLE
$XONSH_TRACEBACK_LOGFILE fails with '~' included

### ISSUE THREAD

**Issue by lazmol** [xonsh/xonsh#2026]
Hello,

firstly thanks for xonsh, I've just started using it, so chances are high I am doing something wrong. Nevertheless it seems to me that $XONSH_TRACEBACK_LOGFILE can not take string like this:
'~/xonsh.log'
It throws an error:
xonsh: $XONSH_TRACEBACK_LOGFILE must be a filepath pointing to a file that either exists and is writable or that can be created.

The value of it is None.
But this works:
log_path = os.path.expanduser(os.path.normpath('~/xonsh.log'))
if not os.path.isfile(log_path):
    open(log_path, 'w').close()
$XONSH_TRACEBACK_LOGFILE = log_path

It is really minor stuff but maybe you could include the expanduser to avoid such problems.
It was on my windows machine.

Cheers
Laszlo

**Comment by gforsyth** [comment]
Hi @lazmol, thanks for reporting! 
We're definitely missing an expand user somewhere in there. 
It looks like we need a call to `xonsh.tools.expandpath` in `xonsh.tools.is_writable_file` if you'd like to try tackling it!

**Comment by scopatz** [comment]
A PR would be most welcome!

**Comment by aweltsch** [comment]
Hi,
I would really like to contribute a patch addressing this issue. @gforsyth has already given quite some details on what to do, but while reading through the source code I stumbled upon `xonsh.built_ins.expand_path` which seems to do something pretty similar to `xonsh.tools.is_writable_file` except for this additional part:
```
# expand ~ according to Bash unquoted rules "Each variable assignment is
# checked for unquoted tilde-prefixes immediately following a ':' or the
# first '='". See the following for more details.
# https://www.gnu.org/software/bash/manual/html_node/Tilde-Expansion.html
```
Is there any reason to prefer `tools.expandpath` over `built_ins.expand_path`?

**Comment by astronouth7303** [comment]
I believe we've had this bug a few times, maybe we should have a generic solution for not-list path environment variables? (List ones have `EnvPath` and friends.)

**Comment by scopatz** [comment]
I believe that the one in `built_ins` is the more general and correct solution.  If we want to make that canonical solution, we probably should move it to tools.

**Comment by aweltsch** [comment]
So which one solution do we want to implement?
If we move the implementation from `built_ins` to `tools` we also need to consider the different behaviour considering tilde expansion (in which cases it's done)

`built_ins.expand_path`: always expands tilde
`tools.expandpath`: expands tilde iff EXPAND_ENV_VARS is `True`

There are some tests (so I guess there is also some code) that relies on the behaviour of `tools.expandpath`. To preserve the old behaviour

**Comment by aweltsch** [comment]
So do we choose to implement @astronouth7303 's propostion or @scopatz 's?

I'm currently working on moving `built_ins.expand_path` -> `tools.expand_path` (aka what @scopat proposed).
There is a small difference in the expansion behaviour of `tools.expandpath` and `builtins.expand_path`
* `tools.expandpath`: expands tilde iff EXPAND_ENV_VARS is set
* `built_ins.expand_path`: always expands tilde
There are some tests that depend on `tools.expandpath` behaviour, so we have to preserve to preserve this.

To finalize I obviously want to write a regression test for this bug. I've got the following problem:
`tools.is_logfile_opt` calls `tools.is_writable_file` which checks if you can write a file. So if the test extends `'~/logfile'` to `'/home/whatever/logfile'` it might of course happen that the file is not writable. Can you think of any way to solve this issue?

**Comment by scopatz** [comment]
> it might of course happen that the file is not writable. Can you think of any way to solve this issue?

You can write to a file in the tests directory or a temp file.  That is the cleanest way to ensure that a file is writable in the tests. Sorry for the delay in the response.

**Comment by mitnk** [comment]
This issue should have been addressed by #2100. Closing.
